arch/x86/core/early_serial.c

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <generated_dts_board.h>
+
+/* Super-primitive 8250/16550 serial output-only driver, 115200 8n1 */
+
+#define _PORT DT_UART_NS16550_PORT_0_BASE_ADDR
+
+static inline void ioport_out8(unsigned short port, unsigned char b)
+{
+	__asm__ volatile("outb %0, %1;\n\t" : : "a"(b), "d"(port));
+}
+
+static inline unsigned char ioport_in8(unsigned short port)
+{
+	unsigned char ret;
+
+	__asm__ volatile("inb %1, %0;\n\t" : "=a"(ret) : "d"(port));
+	return ret;
+}
+
+static void serout(int c)
+{
+	while (!(ioport_in8(_PORT + 5) & 0x20)) {
+	}
+	ioport_out8(_PORT, c);
+}
+
+static int console_out(int c)
+{
+	if (c == '\n') {
+		serout('\r');
+	}
+	serout(c);
+	return c;
+}
+
+extern void __stdout_hook_install(int (*hook)(int));
+extern void __printk_hook_install(int (*fn)(int));
+
+void z_x86_early_serial_init(void)
+{
+	/* In fact Qemu already has most of this set up and works by
+	 * default
+	 */
+	ioport_out8(_PORT+1, 0);    /* IER = 0 */
+	ioport_out8(_PORT+3, 0x80); /* LCR = 8n1 + DLAB select */
+	ioport_out8(_PORT,   1);    /* Divisor Latch low byte */
+	ioport_out8(_PORT+1, 0);    /* Divisor Latch high byte */
+	ioport_out8(_PORT+3, 0x03); /* LCR = 8n1 + DLAB off */
+	ioport_out8(_PORT+4, 0x03); /* MCR = DTR & RTS asserted */
+
+	/* Will be replaced later when a real serial driver comes up */
+	__stdout_hook_install(console_out);
+	__printk_hook_install(console_out);
```

Comment:
Doesn't stdout go to printk() by default?  I swear it did that already.  If not, it probably should.

Reply:
think this is only for printf, i'll remove it